File: Plugins/Kingbotk/AWB Plugin/WikiProjects/WPAustralia.vb
Friend NotInheritable Class WPAustralia
    Inherits PluginBase

    Friend Sub New()
        MyBase.New("") ' Specify alternate names only

        OurSettingsControl = New GenericWithWorkgroups(PluginName, Prefix, True, params)
    End Sub

    Const Prefix As String = "Aus"
    Const PluginName As String = "WikiProject Australia"

    Const PlacesGroup As String = "Places"
    Const SportsGroup As String = "Sports"
    Const OtherGroup As String = "Other topics"

    Dim params() As TemplateParameters =
    {
       New TemplateParameters() With {.StorageKey = "Place", .Group = PlacesGroup, .ParamName = "place"}, _
       New TemplateParameters() With {.StorageKey = "Adel", .Group = PlacesGroup, .ParamName = "Adelaide"}, _
       New TemplateParameters() With {.StorageKey = "Bris", .Group = PlacesGroup, .ParamName = "Brisbane"}, _
       New TemplateParameters() With {.StorageKey = "Canb", .Group = PlacesGroup, .ParamName = "Canberra"}, _
       New TemplateParameters() With {.StorageKey = "Gee", .Group = PlacesGroup, .ParamName = "Geelong"}, _
       New TemplateParameters() With {.StorageKey = "Melb", .Group = PlacesGroup, .ParamName = "Melbourne"}, _
       New TemplateParameters() With {.StorageKey = "Perth", .Group = PlacesGroup, .ParamName = "Perth"}, _
       New TemplateParameters() With {.StorageKey = "Sydney", .Group = PlacesGroup, .ParamName = "Sydney"}, _
       New TemplateParameters() With {.StorageKey = "Sport", .Group = SportsGroup, .ParamName = "Sport"}, _
       New TemplateParameters() With {.StorageKey = "AFL", .Group = SportsGroup, .ParamName = "AFL"}, _
       New TemplateParameters() With {.StorageKey = "NBL", .Group = SportsGroup, .ParamName = "NBL"}, _
       New TemplateParameters() With {.StorageKey = "NRL", .Group = SportsGroup, .ParamName = "NRL"}, _
       New TemplateParameters() With {.StorageKey = "V8", .Group = SportsGroup, .ParamName = "V8"}, _
       New TemplateParameters() With {.StorageKey = "Crime", .Group = OtherGroup, .ParamName = "Crime"}, _
       New TemplateParameters() With {.StorageKey = "Law", .Group = OtherGroup, .ParamName = "Law"}, _
       New TemplateParameters() With {.StorageKey = "Military", .Group = OtherGroup, .ParamName = "Military"}, _
       New TemplateParameters() With {.StorageKey = "Music", .Group = OtherGroup, .ParamName = "Music"}, _
       New TemplateParameters() With {.StorageKey = "Politics", .Group = OtherGroup, .ParamName = "Politics"}
    }

    ' Settings:
    Private OurTab As New TabPage("Australia")
    Private OurSettingsControl As GenericWithWorkgroups

    Protected Friend Overrides ReadOnly Property PluginShortName() As String
        Get
            Return "Australia"
        End Get
    End Property

    Protected Overrides Sub ImportanceParameter(ByVal Importance As Importance)
        Template.NewOrReplaceTemplateParm("importance", Importance.ToString, Me.Article, False, False)
    End Sub
    Protected Friend Overrides ReadOnly Property GenericSettings() As IGenericSettings
        Get
            Return OurSettingsControl
        End Get
    End Property
    Protected Overrides ReadOnly Property CategoryTalkClassParm() As String
        Get
            Return "cat"
        End Get
    End Property
    Protected Overrides ReadOnly Property TemplateTalkClassParm() As String
        Get
            Return "NA"
        End Get
    End Property

    Protected Overrides ReadOnly Property PreferredTemplateName() As String
        Get
            Return PluginName
        End Get
    End Property
    'Protected Overrides ReadOnly Property PreferredTemplateNameRegexString() As String
    '    Get
    '        Return "^[Ww]P Australia$"
    '    End Get
    'End Property

    ' Initialisation:

    Protected Friend Overrides Sub Initialise()
        OurMenuItem = New ToolStripMenuItem("Australia Plugin")
        MyBase.InitialiseBase() ' must set menu item object first
        OurTab.UseVisualStyleBackColor = True
        OurTab.Controls.Add(OurSettingsControl)
    End Sub

    ' Article processing:
    Protected Overrides Function SkipIfContains() As Boolean
        ' None
    End Function
    Protected Overrides Sub ProcessArticleFinish()
        StubClass()
        With OurSettingsControl
            For Each lvi As ListViewItem In .ListView1.Items
                If lvi.Checked Then
                    Dim tp As TemplateParameters = DirectCast(lvi.Tag, TemplateParameters)
                    AddAndLogNewParamWithAYesValue(tp.ParamName.ToLower()) 'Probably needs some reformatting
                End If
            Next
        End With
    End Sub
    Protected Overrides Function TemplateFound() As Boolean
        If CheckForDoublyNamedParameters("V8", "v8") Then Return True ' tag is bad, exit
        If CheckForDoublyNamedParameters("nbl", "NBL") Then Return True
    End Function
    Protected Overrides Sub GotTemplateNotPreferredName(ByVal TemplateName As String)
    End Sub
    Protected Overrides Function WriteTemplateHeader() As String
        Return "{{" & PluginName & _
           Microsoft.VisualBasic.vbCrLf & WriteOutParameterToHeader("class") & _
           WriteOutParameterToHeader("importance")
    End Function

    'User interface:
    Protected Overrides Sub ShowHideOurObjects(ByVal Visible As Boolean)
        PluginManager.ShowHidePluginTab(OurTab, Visible)
    End Sub

    ' XML settings:
    Protected Friend Overrides Sub ReadXML(ByVal Reader As System.Xml.XmlTextReader)
        Dim blnNewVal As Boolean = PluginManager.XMLReadBoolean(Reader, Prefix & "Enabled", Enabled)
        If Not blnNewVal = Enabled Then Enabled = blnNewVal ' Mustn't set if the same or we get extra tabs
        OurSettingsControl.ReadXML(Reader)
    End Sub
    Protected Friend Overrides Sub Reset()
        OurSettingsControl.Reset()
    End Sub
    Protected Friend Overrides Sub WriteXML(ByVal Writer As System.Xml.XmlTextWriter)
        Writer.WriteAttributeString(Prefix & "Enabled", Enabled.ToString)
        OurSettingsControl.WriteXML(Writer)
    End Sub

    ' Not implemented:
    Friend Overrides ReadOnly Property HasReqPhotoParam() As Boolean
        Get
            Return False
        End Get
    End Property
    Friend Overrides Sub ReqPhoto()
    End Sub
End Class

File: Plugins/Kingbotk Plugin/WikiProjects/WPAustralia.cs
/*
Copyright (C) 2008 Stephen Kennedy (Kingboyk) http://www.sdk-software.com/
Copyright (C) 2008 Sam Reed (Reedy) http://www.reedyboy.net/

This program is free software; you can redistribute it and/or modify it under the terms of Version 2 of the GNU General Public License as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License Version 2 along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

using System.Windows.Forms;
using System.Xml;

namespace AutoWikiBrowser.Plugins.Kingbotk.WikiProjects
{
    internal sealed class WPAustralia : PluginBase
    {
        internal WPAustralia() : base("WPAUSTRALIA|WP Australia|WPAUS|WPAustralia")
        {
            // Specify alternate names only
            _ourSettingsControl = new GenericWithWorkgroups(PluginName, Prefix, true, _params);
        }

        private const string Prefix = "Aus";

        private const string PluginName = "WikiProject Australia";
        private const string PlacesGroup = "Places";
        private const string SportsGroup = "Sports";

        private const string OtherGroup = "Other topics";

        private readonly TemplateParameters[] _params =
        {
            new TemplateParameters
            {
                StorageKey = "Place",
                Group = PlacesGroup,
                ParamName = "place"
            },
            new TemplateParameters
            {
                StorageKey = "AAT",
                Group = PlacesGroup,
                ParamName = "AAT"
            },
            new TemplateParameters
            {
                StorageKey = "Adel",
                Group = PlacesGroup,
                ParamName = "Adelaide"
            },
            new TemplateParameters
            {
                StorageKey = "Bris",
                Group = PlacesGroup,
                ParamName = "Brisbane"
            },
            new TemplateParameters
            {
                StorageKey = "Canb",
                Group = PlacesGroup,
                ParamName = "Canberra"
            },
            new TemplateParameters
            {
                StorageKey = "Gee",
                Group = PlacesGroup,
                ParamName = "Geelong"
            },
            new TemplateParameters
            {
                StorageKey = "Melb",
                Group = PlacesGroup,
                ParamName = "Melbourne"
            },
            new TemplateParameters
            {
                StorageKey = "Noongar",
                Group = PlacesGroup,
                ParamName = "Noongar"
            },
            new TemplateParameters
            {
                StorageKey = "Perth",
                Group = PlacesGroup,
                ParamName = "Perth"
            },
            new TemplateParameters
            {
                StorageKey = "Sydney",
                Group = PlacesGroup,
                ParamName = "Sydney"
            },
            new TemplateParameters
            {
                StorageKey = "River",
                Group = PlacesGroup,
                ParamName = "Riverina"
            },
            new TemplateParameters
            {
                StorageKey = "TAS",
                Group = PlacesGroup,
                ParamName = "TAS"
            },
            new TemplateParameters
            {
                StorageKey = "Sport",
                Group = SportsGroup,
                ParamName = "sports"
            },
            new TemplateParameters
            {
                StorageKey = "AFL",
                Group = SportsGroup,
                ParamName = "afl"
            },
            new TemplateParameters
            {
                StorageKey = "NBL",
                Group = SportsGroup,
                ParamName = "nbl"
            },
            new TemplateParameters
            {
                StorageKey = "NRL",
                Group = SportsGroup,
                ParamName = "nrl"
            },
            new TemplateParameters
            {
                StorageKey = "V8",
                Group = SportsGroup,
                ParamName = "v8"
            },
            new TemplateParameters
            {
                StorageKey = "Crime",
                Group = OtherGroup,
                ParamName = "crime"
            },
            new TemplateParameters
            {
                StorageKey = "Law",
                Group = OtherGroup,
                ParamName = "law"
            },
            new TemplateParameters
            {
                StorageKey = "Military",
                Group = OtherGroup,
                ParamName = "military"
            },
            new TemplateParameters
            {
                StorageKey = "Music",
                Group = OtherGroup,
                ParamName = "music"
            },
            new TemplateParameters
            {
                StorageKey = "Politics",
                Group = OtherGroup,
                ParamName = "politics"
            }
        };

        // Settings:
        private readonly TabPage _ourTab = new TabPage("Australia");

        private readonly GenericWithWorkgroups _ourSettingsControl;

        protected internal override string PluginShortName
        {
            get { return "Australia"; }
        }

        protected override void ImportanceParameter(Importance importance)
        {
            Template.NewOrReplaceTemplateParm("importance", importance.ToString(), TheArticle, false, false);
        }

        protected internal override IGenericSettings GenericSettings
        {
            get { return _ourSettingsControl; }
        }

        protected override string PreferredTemplateName
        {
            get { return PluginName; }
        }

        // Initialisation:

        protected internal override void Initialise()
        {
            OurMenuItem = new ToolStripMenuItem("Australia Plugin");
            InitialiseBase();
            // must set menu item object first
            _ourTab.UseVisualStyleBackColor = true;
            _ourTab.Controls.Add(_ourSettingsControl);
        }

        // Article processing:
        protected override bool SkipIfContains()
        {
            // None
            return false;
        }

        protected override void ProcessArticleFinish()
        {
            StubClass();
            foreach (ListViewItem lvi in _ourSettingsControl.ListView1.Items)
            {
                if (lvi.Checked)
                {
                    TemplateParameters tp = (TemplateParameters) lvi.Tag;
                    AddAndLogNewParamWithAYesValue(tp.ParamName.ToLower());
                    // Probably needs some reformatting
                }
            }
        }

        protected override bool TemplateFound()
        {
            // Nothing to do here
            return false;
        }

        protected override string WriteTemplateHeader()
        {
            return "{{" + PluginName + WriteOutParameterToHeader("class") + WriteOutParameterToHeader("importance");
        }

        // User interface:
        protected override void ShowHideOurObjects(bool visible)
        {
            PluginManager.ShowHidePluginTab(_ourTab, visible);
        }

        // XML settings:
        protected internal override void ReadXML(XmlTextReader reader)
        {
            Enabled = PluginManager.XMLReadBoolean(reader, Prefix + "Enabled", Enabled);
            _ourSettingsControl.ReadXML(reader);
        }

        protected internal override void Reset()
        {
            _ourSettingsControl.Reset();
        }

        protected internal override void WriteXML(XmlTextWriter writer)
        {
            writer.WriteAttributeString(Prefix + "Enabled", Enabled.ToString());
            _ourSettingsControl.WriteXML(writer);
        }

        // Not implemented:
        internal override bool HasReqPhotoParam
        {
            get { return false; }
        }

        internal override void ReqPhoto()
        {
        }
    }
}
